src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -1100,37 +1150,198 @@ private void PopulateParamsFromRestRequest(Dictionary<string, object?> parameter
                         );
             }
 
-            if (!isAuthorized)
+            return isAuthorized;
+        }
+
+        /// <summary>
+        /// Performs authorization checks on entity level permissions and field level permissions for every entity and field
+        /// referenced in a GraphQL mutation for the given client role.
+        /// </summary>
+        /// <param name="context">Middleware context.</param>
+        /// <param name="clientRole">Client role header value extracted from the middleware context of the mutation</param>
+        /// <param name="entityName">Top level entity name.</param>
+        /// <param name="operation">Mutation operation</param>
+        /// <param name="inputArgumentName">Name of the input argument (differs based on point/multiple mutation).</param>
+        /// <param name="parametersDictionary">Dictionary of key/value pairs for the argument name/value.</param>
+        /// <exception cref="DataApiBuilderException">Throws exception when an authorization check fails.</exception>
+        private void AuthorizeEntityAndFieldsForMutation(
+            IMiddlewareContext context,
+            string clientRole,
+            string entityName,
+            EntityActionOperation operation,
+            string inputArgumentName,
+            IDictionary<string, object?> parametersDictionary
+            )
+        {
+            IInputField schemaForArgument = context.Selection.Field.Arguments[inputArgumentName];
+
+            // Dictionary to store a mapping from entityName to all field names being referenced from that entity in the mutation.
+            Dictionary<string, HashSet<string>> entityToFields = new();
+            object? parameters;
+            if (parametersDictionary.TryGetValue(inputArgumentName, out parameters))
+            {
+                // Get all the entity names and field names referenced in the mutation.
+                PopulateMutationEntityAndFieldsToAuthorize(entityToFields, schemaForArgument, entityName, context, parameters);
+            }
+
+            // Perform authorization checks at field level.
+            foreach ((string entityNameInMutation, HashSet<string> fieldsInEntity) in entityToFields)
```

Comment:
Seems like there's a case where `entityToFields` can be empty? e.g. when `if (parametersDictionary.TryGetValue(inputArgumentName, out parameters))` returns false?

Reply:
No, the schema is generated such that parameters would never be null. I'll anyways add an else block to throw an exception.